### PR TITLE
IPAM Subnet Resource and Data Source

### PR DESCRIPTION
**Pull request by mathewab**
- Added resource and data source implementation for ipam/subnet

### CODE REVIEW

File: docs/data-sources/ipam_subnets.md [hunk 1]
```diff
@@ -0,0 +1,773 @@
...
+
+<a id="nestedatt--results--inheritance_sources--ddns_hostname_block"></a>
+### Nested Schema for `results.inheritance_sources.ddns_hostname_block`
+
+Optional:
+
+- `action` (String) The inheritance setting.  Valid values are: * _inherit_: Use the inherited value. * _override_: Use the value set in the object.  Defaults to _inherit_.
+- `source` (String) The resource identifier.
+- `value` (Attributes) (see [below for nested schema](#nestedatt--results--inheritance_sources--ddns_hostname_block--value))
+
```
Comment (venkat-iblox) [inline docs/data-sources/ipam_subnets.md:359]: We should start adding the support for querying inheritance in the read/list API for objects that support Inheritance

Comment (mathewab) [reply docs/data-sources/ipam_subnets.md:359]: We have a different story/bug for it. We can pick it up soon.

File: internal/service/ipam/api_subnet_data_source.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	client, ok := req.ProviderData.(*bloxoneclient.APIClient)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected DataSource Configure Type",
+			fmt.Sprintf("Expected *bloxoneclient.APIClient, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
```
Comment (venkat-iblox) [inline internal/service/ipam/api_subnet_data_source.go:86]: Don't we repeat these lines in every datasource and same for resources too. We should probably move it to a common package for all to use

File: internal/service/ipam/model_ipamsvc_dhcp_config.go [hunk 3]
```diff
@@ -47,13 +46,11 @@ var IpamsvcDHCPConfigResourceSchemaAttributes = map[string]schema.Attribute{
 	"abandoned_reclaim_time": schema.Int64Attribute{
 		Optional:            true,
 		Computed:            true,
-		Default:             int64default.StaticInt64(3600),
```
Comment (venkat-iblox) [inline internal/service/ipam/model_ipamsvc_dhcp_config.go:50]: Any reason why this had to be removed?

File: internal/service/ipam/api_subnet_data_source.go (hunk 2 above)
Comment (mathewab) [reply internal/service/ipam/api_subnet_data_source.go:86]: Yes, I am sure there might be more to restructure. We can maybe focus on it together as a story, and do it once all the objects are done. Right now this is generated code, so it isn't much work anyway.

File: internal/service/ipam/model_ipamsvc_dhcp_config.go (hunk 3 above)
Comment (mathewab) [reply internal/service/ipam/model_ipamsvc_dhcp_config.go:50]: The `abandoned_reclaim_time` and `abandoned_reclaim_time_v6` are not supported for subnet. Having the default there causes an error in API, as the API doesn't expect the value in the field.

File: internal/service/ipam/api_subnet_data_source.go (hunk 2 above)
Comment (venkat-iblox) [reply internal/service/ipam/api_subnet_data_source.go:86]: Yes, thats the idea. I don't expect you to do it in this PR. Just pointing it out

**Review by venkat-iblox — APPROVED** [review]

**Pull request merged**